server/src/weaverbird/backends/pandas_executor/steps/utils/formula.py

```diff
@@ -1,40 +1,21 @@
-import re
-
 import numpy as np
 from pandas import DataFrame
 
-COLUMN_PATTERN = re.compile(r'(\`[^\`]*\`)')
-
-
-# In pandas, columns containing spaces must be escaped with backticks:
-def clean_formula_element(elem: str) -> str:
-    if COLUMN_PATTERN.match(elem):
-        # this is a column name: return it as is
-        return elem
-    # - forbid '=' char
-    # - replace [ and ] with `
-    #   ([] is mongo's syntax to escape columns containing spaces)
-    return elem.replace('=', '').replace('[', '`').replace(']', '`')
-
+from weaverbird.pipeline.formula_ast.eval import FormulaParser
+from weaverbird.pipeline.formula_ast.types import Expression, format_expr
 
-def clean_formula(formula: str) -> str:
-    """
-    Translate mongo's syntax to hande columns names containing spaces to pandas syntax.
 
-    Example:
-        >>> clean_formula('colA * `col B` * [col C] * `[col D]`')
-        'colA * `col B` * `col C` * `[col D]`'
-    """
-    formula_splitted = COLUMN_PATTERN.split(formula)
-    return ''.join(clean_formula_element(elem) for elem in formula_splitted)
+def formula_expression_to_pandas(tree: Expression) -> str | int | bool | float:
+    return format_expr(tree, column_start_seq='`', column_end_seq='`', bools_as_py=True)
 
 
 def eval_formula(df: DataFrame, formula: str) -> DataFrame:
+    pandas_expr = formula_expression_to_pandas(FormulaParser(formula).parse())
     try:
-        result = df.eval(formula)
+        result = df.eval(pandas_expr)
     except Exception:
         # for all cases not handled by NumExpr
-        result = df.eval(formula, engine='python')
+        result = df.eval(pandas_expr, engine='python')
```

Comment:
One of the interest of parsing the formula ourselves is to overcome the limitations of pandas' eval and use standard vectored operations instead.
So it would be best if we did the eval_formula was doing operations directly when reading a node:
- walking along a name node => df[that name]
- walking along a constant node => that constant value
- walking along an operation node => getattr(right_result,'__<this op>__')(left_result) (maybe there's a more idiomatic way of writing this, or just doing a match with the 6 cases would be enough)

Reply:
I assumed pandas would be faster than us, that's why I went with it although we could maybe build a function that'd be applied only once.

But since we now control the AST, I think `numexpr` covers all our usecases: https://numexpr.readthedocs.io/projects/NumExpr3/en/latest/user_guide.html#supported-operators . Maybe we could build a numexpr-compatible formula from the AST and remove the `try/except` ? It'll probably be much faster than us (no python function call stack/dereferencing).

Again, I'm also fine with the manual eval approach, whatever the majority prefers

Reply:
I remember a case where we had a problem with something like `df.eval('a + b')`, and where `df['a'] + df['b']` worked.
I thunk this was because the nullable dtypes were not supported in eval. I'll try to find this bug again in the archives

Reply:
Ah shit forgot about connectors and their exotic dtypes, nice catch :+1:  Okay let's go for manual eval then. I'll see what I can do to manipulate `Series` as much as possible rather than `DataFrame`s

Reply:
I also found this issue about it: https://github.com/ToucanToco/weaverbird/issues/1294

Reply:
Pandas manual eval implemented in https://github.com/ToucanToco/weaverbird/commit/b32a026fb6598f5040d1807092d536c0b6fa17dd